File: desktop/src/alStatus.ts
import catalog from "../../src/ethercat_debug_tool/protocol/al_status_codes.json";
import { englishAlStatusNames } from "./alStatusEnglish";

export type AlStatusLanguage = "zh" | "en";

export interface AlStatusInfo {
  name: string;
  detail: string;
  action: string;
  known: boolean;
}

const codes = catalog.codes as Record<string, Omit<AlStatusInfo, "known">>;
const vendorSpecific = catalog.ranges["0x8000-0xFFFF"];
const unknown = catalog.unknown;

export function alStatusInfo(code: number, language: AlStatusLanguage = "zh"): AlStatusInfo {
  const normalized = code & 0xffff;
  const key = `0x${normalized.toString(16).toUpperCase().padStart(4, "0")}`;
  const info = codes[key];
  if (info) {
    if (language === "zh") return { ...info, known: true };
    const name = englishAlStatusNames[key] ?? "Recognized AL status code";
    if (normalized === 0x0050) return {
      name,
      detail: "The SII EEPROM is not assigned to the PDI, or the firmware cannot obtain the EEPROM access it requires.",
      action: "Read 0x0500, 0x0501 and 0x0502 together; inspect ECAT/PDI ownership, Busy/error bits, and the slave firmware state machine.",
      known: true,
    };
    return {
      name,
      detail: normalized === 0 ? "The slave reports no AL status error." : `The slave reported “${name}” while processing its EtherCAT state or configuration.`,
      action: normalized === 0 ? "No action is required." : "Check the current and requested states, AL Status registers, ESI/SII configuration, and the device diagnostic log.",
      known: true,
    };
  }
  if (normalized >= 0x8000) {
    return language === "zh" ? { ...vendorSpecific, known: false } : { name: "Vendor-specific AL status code", detail: "This code is defined by the device vendor, not by the common EtherCAT AL status table.", action: "Consult the device manual, ESI annotations, and vendor diagnostic objects.", known: false };
  }
  return language === "zh" ? { ...unknown, known: false } : { name: "Unlisted AL status code", detail: "This value is not in the current common status table and may be reserved or a newer extension.", action: "Read register 0x0134 and consult the device manual, ESI, and slave log.", known: false };
}

export const standardAlStatusCodeCount = Object.keys(codes).length;

File: src/ethercat_debug_tool/al_status.py
from __future__ import annotations

import json
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path


@dataclass(frozen=True, slots=True)
class AlStatusInfo:
    name: str
    detail: str
    action: str
    known: bool


@lru_cache(maxsize=1)
def _payload() -> dict[str, object]:
    path = Path(__file__).with_name("protocol") / "al_status_codes.json"
    return json.loads(path.read_text(encoding="utf-8"))


@lru_cache(maxsize=1)
def _catalog() -> dict[int, AlStatusInfo]:
    codes = _payload()["codes"]
    assert isinstance(codes, dict)
    return {
        int(code, 16): AlStatusInfo(
            name=str(value["name"]),
            detail=str(value["detail"]),
            action=str(value["action"]),
            known=True,
        )
        for code, value in codes.items()
    }


def _fallback(section: str, key: str | None = None) -> AlStatusInfo:
    value = _payload()[section]
    if key is not None:
        assert isinstance(value, dict)
        value = value[key]
    assert isinstance(value, dict)
    return AlStatusInfo(str(value["name"]), str(value["detail"]), str(value["action"]), False)


def al_status_info(code: int) -> AlStatusInfo:
    normalized = code & 0xFFFF
    if info := _catalog().get(normalized):
        return info
    if normalized >= 0x8000:
        return _fallback("ranges", "0x8000-0xFFFF")
    return _fallback("unknown")
